src/Generator/Generators/CSharp/CSharpMarshal.cs

```diff
@@ -382,16 +385,26 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
             if (!VisitType(array, quals))
                 return false;
 
+            Class @class;
             switch (array.SizeType)
             {
                 case ArrayType.ArraySize.Constant:
                     var supportBefore = Context.SupportBefore;
                     supportBefore.WriteLine("if ({0} != null)", Context.ArgName);
                     supportBefore.WriteStartBraceIndent();
+                    if (array.Type.Desugar().TryGetClass(out @class) && @class.IsRefType)
+                    {
+                        supportBefore.WriteLine("if (value.Length != {0})", array.Size);
+                        supportBefore.WriteLineIndent("throw new ArrayTypeMismatchException(\"The provided array's dimensions doesn't match the required size.\");");
```

Comment:
Hmmm. That is okay. What I want to say that suppose (like the test) the array is of size 4 and the user gave array of size 2. Then the indexes 0 & 1 are all within range. But exception will be thrown.